packages/hardhat-core/src/internal/artifacts.ts

```diff
@@ -867,6 +886,235 @@ Please replace "${contractName}" for the correct contract name wherever you are
+      )
+    ).flat();
+  }
+
+  public async getBuildInfoPaths(): Promise<string[]> {
+    return (
+      await Promise.all(
+        this._sourcesInPriorityOrder.map((s) => s.getBuildInfoPaths())
+      )
+    ).flat();
+  }
+
+  public clearCache(): void {
+    for (const source of this._sourcesInPriorityOrder) {
+      source.clearCache();
+    }
+  }
+
+  public disableCache(): void {
+    for (const source of this._sourcesInPriorityOrder) {
+      source.disableCache();
+    }
+  }
+
+  public saveArtifactAndDebugFile(
+    artifact: Artifact,
+    pathToBuildInfo?: string
+  ): Promise<void> {
+    return this._sourcesInPriorityOrder[0].saveArtifactAndDebugFile(
+      artifact,
+      pathToBuildInfo
+    );
+  }
+
+  public saveBuildInfo(
+    solcVersion: string,
+    solcLongVersion: string,
+    input: CompilerInput,
+    output: CompilerOutput
+  ): Promise<string> {
+    return this._sourcesInPriorityOrder[0].saveBuildInfo(
+      solcVersion,
+      solcLongVersion,
+      input,
+      output
+    );
+  }
+
+  public formArtifactPathFromFullyQualifiedName(
+    fullyQualifiedName: string
+  ): string {
+    return this._sourcesInPriorityOrder[0].formArtifactPathFromFullyQualifiedName(
+      fullyQualifiedName
+    );
+  }
+
+  public addValidArtifacts(
+    validArtifacts: Array<{ sourceName: string; artifacts: string[] }>
+  ) {
+    this._sourcesInPriorityOrder[0].addValidArtifacts(validArtifacts);
+  }
+
+  public async removeObsoleteArtifacts() {
+    this._sourcesInPriorityOrder[0].removeObsoleteArtifacts();
+  }
+
+  private async _getFirstValueFromSources(
+    method: "readArtifact",
+    key: string
+  ): Promise<Artifact>;
+
+  private async _getFirstValueFromSources(
+    method: "artifactExists",
+    key: string
+  ): Promise<boolean>;
+
+  private async _getFirstValueFromSources(
+    method: "getBuildInfo",
+    key: string
+  ): Promise<BuildInfo | undefined>;
+
+  private async _getFirstValueFromSources(
+    method: keyof ArtifactSource,
+    key: string
+  ): Promise<boolean | Artifact | string[] | BuildInfo | undefined> {
+    /* iterate over the sources to resolve the given name. if a source returns
+     * undefined or NOT_FOUND, continue on to the next source. preserve the
+     * latest thrown error so that its helpful error message can be delivered
+     * in the case where latter sources simply return undefined. if all sources
+     * are exhausted, and still nothing is found, throw NOT_FOUND. */
+    let caughtError: unknown | undefined;
+    for (const source of this._sourcesInPriorityOrder) {
+      try {
+        const value = await source[method](key);
+        if (value === undefined) {
+          continue;
+        }
+        return value;
+      } catch (error) {
+        if (
+          error instanceof HardhatError &&
+          error.number === ERRORS.ARTIFACTS.NOT_FOUND.number
+        ) {
+          caughtError = error;
```

Comment:
I think it would be better to modify the underlying logic to always return undefined and never throw, and remove this try/catch. The result would be the same as what's going on here, but much clearer.

Reply:
I would love to see this get cleaned up / simplified, but there's a good reason to catch and re-throw here: that lower layer offers typo suggestions in the error, which we can't replicate at this level. (at least, not without exposing more stuff on the interface

Reply:
(Resolving, but see my other comment about this)